File: core/needle_detector.py
"""
Phát hiện đầu kim châm cứu được sơn màu xanh lá.
Module này độc lập với UI — dùng được trên desktop lẫn mobile.
"""

import cv2
import numpy as np
import imutils
import math


# ── Thông số camera calibration (Raspberry Pi webcam gốc) ──────────────────
# Khi dùng camera điện thoại: bỏ qua undistort hoặc calibrate lại
_CAMERA_MATRIX = np.array([
    [936.37183128,   0.0,         493.07464865],
    [  0.0,         937.43778885,  93.18820841],
    [  0.0,           0.0,           1.0      ]
])
_DIST_COEFFS = np.array([[0.0171085, -0.18085887, 0.01305453, -0.01131838, 0.20937844]])

# ── Thông số lọc màu HSV (xanh lá) ────────────────────────────────────────
# Điều chỉnh nếu camera điện thoại cho ra màu sắc khác
HSV_LOWER = np.array([97,  73, 140])
HSV_UPPER = np.array([136, 255, 255])

# Diện tích contour tối thiểu (pixel²) để xác nhận là đầu kim
MIN_CONTOUR_AREA = 70

_KERNEL = np.ones((3, 3), np.uint8)


class DetectionResult:
    """Kết quả nhận diện cho 1 khung hình."""

    def __init__(self):
        self.found: bool = False
        self.center: tuple = (0, 0)   # (cx, cy) pixel
        self.angle: int = 0           # Degrees 0–90
        self.radian: float = 0.0      # Radians
        self.area: float = 0.0        # Diện tích contour
        self.annotated_frame = None   # Frame BGR đã vẽ bounding box

    def __repr__(self):
        if self.found:
            return f"<DetectionResult found center={self.center} angle={self.angle}°>"
        return "<DetectionResult not_found>"


class NeedleDetector:
    """
    Phát hiện đầu kim châm cứu sơn xanh lá trong 1 khung hình.

    Sử dụng:
        detector = NeedleDetector()
        result = detector.detect(frame)
        if result.found:
            print(result.center, result.angle)
    """

    def __init__(self,
                 use_calibration: bool = True,
                 hsv_lower=None,
                 hsv_upper=None,
                 min_area: int = MIN_CONTOUR_AREA):
        """
        Args:
            use_calibration: Dùng undistort hay không.
                             False khi dùng camera điện thoại chưa calibrate.
            hsv_lower: Ghi đè ngưỡng HSV dưới.
            hsv_upper: Ghi đè ngưỡng HSV trên.
            min_area:  Diện tích tối thiểu để nhận dạng là đầu kim.
        """
        self.use_calibration = use_calibration
        self.lower = hsv_lower if hsv_lower is not None else HSV_LOWER.copy()
        self.upper = hsv_upper if hsv_upper is not None else HSV_UPPER.copy()
        self.min_area = min_area

    # ── Public API ──────────────────────────────────────────────────────────

    def detect(self, frame) -> DetectionResult:
        """
        Nhận diện đầu kim trong frame BGR.

        Trả về DetectionResult (luôn có annotated_frame dù không tìm thấy kim).
        """
        result = DetectionResult()

        dst = self._preprocess(frame)
        result.annotated_frame = dst.copy()

        hsv = cv2.cvtColor(dst, cv2.COLOR_BGR2HSV)
        mask = cv2.inRange(hsv, self.lower, self.upper)

        contours = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
        contours = imutils.grab_contours(contours)

        best_area = 0
        for c in contours:
            area = cv2.contourArea(c)
            if area < self.min_area or area <= best_area:
                continue
            best_area = area

            rect = cv2.minAreaRect(c)
            box = cv2.boxPoints(rect)
            box = np.int0(box)

            cx = int(rect[0][0])
            cy = int(rect[0][1])
            w  = int(rect[1][0])
            h  = int(rect[1][1])
            ang = int(rect[2])

            # Chuẩn hóa góc về 0–90° (không phân biệt nghiêng trái/phải)
            if w < h:
                ang = 90 - ang

            cv2.drawContours(result.annotated_frame, [box], 0, (0, 0, 255), 2)

            result.found  = True
            result.center = (cx, cy)
            result.angle  = ang
            result.radian = math.radians(ang)
            result.area   = area

        return result

    def set_hsv_range(self, lower, upper):
        """Cập nhật ngưỡng màu HSV (dùng khi camera điện thoại cần hiệu chỉnh)."""
        self.lower = np.array(lower)
        self.upper = np.array(upper)

    # ── Internal ────────────────────────────────────────────────────────────

    def _preprocess(self, frame):
        """Undistort (tuỳ chọn) + blur + erode + dilate."""
        if self.use_calibration:
            h, w = frame.shape[:2]
            new_mat, _ = cv2.getOptimalNewCameraMatrix(
                _CAMERA_MATRIX, _DIST_COEFFS, (w, h), 1, (w, h)
            )
            dst = cv2.undistort(frame, _CAMERA_MATRIX, _DIST_COEFFS, None, new_mat)
        else:
            dst = frame.copy()

        blurred = cv2.blur(dst, (3, 3))
        eroded  = cv2.erode(blurred, (3, 3))
        dilated = cv2.dilate(eroded, _KERNEL, iterations=3)
        return dilated

File: web/js/needle_detector.js
'use strict';

/**
 * Phát hiện đầu kim châm cứu sơn màu xanh lá bằng OpenCV.js.
 * Port từ core/needle_detector.py — logic HSV hoàn toàn giống nhau.
 *
 * Lưu ý kênh màu:
 *   Canvas → RGBA → cv.COLOR_RGBA2BGR → BGR → cv.COLOR_BGR2HSV → HSV
 *   Ngưỡng HSV giữ nguyên từ code Python gốc.
 */

class NeedleDetector {
  /**
   * @param {object} opts
   * @param {number[]} opts.hsvLower      - [H, S, V] ngưỡng dưới (default từ code gốc)
   * @param {number[]} opts.hsvUpper      - [H, S, V] ngưỡng trên
   * @param {number}   opts.minArea       - diện tích contour tối thiểu (pixel²)
   * @param {number}   opts.dilateIter    - số lần dilate
   */
  constructor(opts = {}) {
    this.hsvLower   = opts.hsvLower   ?? [97,  73, 140];
    this.hsvUpper   = opts.hsvUpper   ?? [136, 255, 255];
    this.minArea    = opts.minArea    ?? 70;
    this.dilateIter = opts.dilateIter ?? 3;

    // Mats tái sử dụng (khởi tạo khi frame đầu tiên đến)
    this._kernel    = null;
    this._lowerMat  = null;
    this._upperMat  = null;
    this._lastRows  = 0;
    this._lastCols  = 0;
  }

  /**
   * Phát hiện kim trong 1 khung hình.
   * @param   {HTMLCanvasElement} canvas
   * @returns {{ found, center, angle, radian, area, vertices }}
   */
  detect(canvas) {
    const result = {
      found: false, center: [0, 0],
      angle: 0, radian: 0, area: 0, vertices: null
    };

    // Mats tạm — được giải phóng trong finally
    let src = null, bgr = null, blr = null,
        ero = null, dil = null, hsv = null, mask = null,
        contours = null, hierarchy = null;

    try {
      this._ensureStaticMats(canvas.height, canvas.width);

      src = cv.imread(canvas);                          // RGBA

      bgr = new cv.Mat();
      cv.cvtColor(src, bgr, cv.COLOR_RGBA2BGR);         // RGBA → BGR

      blr = new cv.Mat();
      cv.blur(bgr, blr, new cv.Size(3, 3));             // làm mờ

      ero = new cv.Mat();
      cv.erode(blr, ero, this._kernel);                  // erode

      dil = new cv.Mat();
      cv.dilate(ero, dil, this._kernel,
                new cv.Point(-1, -1), this.dilateIter); // dilate × 3

      hsv = new cv.Mat();
      cv.cvtColor(dil, hsv, cv.COLOR_BGR2HSV);          // BGR → HSV

      mask = new cv.Mat();
      cv.inRange(hsv, this._lowerMat, this._upperMat, mask); // lọc màu

      contours  = new cv.MatVector();
      hierarchy = new cv.Mat();
      cv.findContours(mask, contours, hierarchy,
                      cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);

      let maxArea = 0;
      for (let i = 0; i < contours.size(); i++) {
        const c    = contours.get(i);
        const area = cv.contourArea(c);

        if (area > this.minArea && area > maxArea) {
          maxArea = area;

          const rect = cv.minAreaRect(c);
          let   ang  = Math.round(rect.angle);
          if (rect.size.width < rect.size.height) ang = 90 - ang;

          result.found    = true;
          result.center   = [Math.round(rect.center.x), Math.round(rect.center.y)];
          result.angle    = ang;
          result.radian   = ang * Math.PI / 180;
          result.area     = area;
          result.vertices = NeedleDetector._rectVertices(rect);
        }
        c.delete();
      }

    } catch (e) {
      console.error('NeedleDetector.detect error:', e);
    } finally {
      [src, bgr, blr, ero, dil, hsv, mask, contours, hierarchy]
        .forEach(m => { try { if (m && !m.isDeleted?.()) m.delete(); } catch(_){} });
    }

    return result;
  }

  /** Cập nhật ngưỡng HSV (khi cần điều chỉnh theo camera điện thoại). */
  setHSVRange(lower, upper) {
    this.hsvLower = lower;
    this.hsvUpper = upper;
    this._invalidateMats();
  }

  /** Giải phóng Mats tĩnh khi không dùng nữa. */
  dispose() {
    this._invalidateMats();
    if (this._kernel) { this._kernel.delete(); this._kernel = null; }
  }

  // ── Private ──────────────────────────────────────────────────────────────

  /** Khởi tạo/tái tạo Mats tĩnh khi kích thước frame thay đổi. */
  _ensureStaticMats(rows, cols) {
    if (this._kernel) {
      if (this._lastRows === rows && this._lastCols === cols
          && this._lowerMat) return;   // không thay đổi
    }

    this._invalidateMats();

    this._kernel = cv.Mat.ones(3, 3, cv.CV_8U);

    const [lh, ls, lv] = this.hsvLower;
    const [uh, us, uv] = this.hsvUpper;
    // Tạo Mat kích thước frame để inRange hoạt động ổn định
    this._lowerMat = new cv.Mat(rows, cols, cv.CV_8UC3,
                                new cv.Scalar(lh, ls, lv));
    this._upperMat = new cv.Mat(rows, cols, cv.CV_8UC3,
                                new cv.Scalar(uh, us, uv));
    this._lastRows = rows;
    this._lastCols = cols;
  }

  _invalidateMats() {
    ['_lowerMat', '_upperMat'].forEach(k => {
      try { if (this[k] && !this[k].isDeleted?.()) this[k].delete(); } catch(_){}
      this[k] = null;
    });
  }

  /**
   * Tính 4 góc của RotatedRect từ tâm + kích thước + góc.
   * cv.RotatedRect.points() không ổn định trên mọi phiên bản OpenCV.js,
   * nên tự tính bằng trigonometry.
   */
  static _rectVertices({ center, size, angle }) {
    const rad = angle * Math.PI / 180;
    const cos = Math.cos(rad), sin = Math.sin(rad);
    const hw = size.width / 2, hh = size.height / 2;
    return [
      { x: center.x - cos*hw + sin*hh, y: center.y - sin*hw - cos*hh },
      { x: center.x + cos*hw + sin*hh, y: center.y + sin*hw - cos*hh },
      { x: center.x + cos*hw - sin*hh, y: center.y + sin*hw + cos*hh },
      { x: center.x - cos*hw - sin*hh, y: center.y - sin*hw + cos*hh },
    ];
  }
}
